### ISSUE TITLE
Cannot delete content from "optional" fields

### ISSUE THREAD

**Issue by pointybeard** [symphonycms/symphony-2#2733]
Apologies in advance if this has already been addressed, however, the behaviour is still present in the 2.7.x branch and I've not located a fix or existing issue. Appears this bug has been around since 18th of May.

**Affected Symphony version(s) :** Since PR #2664 

This relates to code in EntryManager line 119 (https://github.com/symphonycms/symphony-2/blob/2.7.x/symphony/lib/toolkit/class.entrymanager.php#L119)

### Expected Behaviour

If there is a field flagged as optional, editing it to have an empty value should cause anything that was there to be deleted.

### Current Behaviour

The original value persists.

### Steps to reproduce

1. Create a new section.
2. Add a text input field to that section and set it to optional
3. Create a new entry in that section, ensuring to add a value to the text input. e.g. "APPLE"
4. Go back to the entry to edit it. Delete the value "APPLE" from the text input field, then save the entry.
5. Observe the value "APPLE" persists.

### Reason

When an entry is saved, field data is passed through this method. For an optional field with no value, the data looks something like this:

```
int(113)
array(0) {
}
```

This code is executed near the top of `EntryManager::saveFieldData()`:

```
        // Check if we have field data
        if (!is_array($field) || empty($field)) {
            return;
        }
```

When this code is executed, it will skip over fields with an empty value. Consider the previous code https://github.com/symphonycms/symphony-2/blob/0dd2c1238a62aa620d7f7e5737843380a31c8b6a/symphony/lib/toolkit/class.entrymanager.php#L186

Note that a call to `Symphony::Database()->delete(...)` is triggered first. This allows the desired behaviour to occur. i.e. removing a value from an optional field. 

### Solution

Delete field data before checking for an empty state.

**Comment by nitriques** [comment]
Hey Alistair ! Great to see you report a bug! I confirm the issue. Thanks!

**Comment by nitriques** [comment]
@michael-e @pointybeard 

Re: https://github.com/symphonycms/symphony-2/commit/257aab2deb688c4b6c68805b102dabd4ba880c7e#commitcomment-23450932

Let's continue here. I have a working fix which restore the idea of deleting when we have no data.

The only thing that would still be different is the case of the `add` method, which did not delete data before inserting. I think that's why it got into the refactored method.

I think it is the right thing do to: we should always clean up first and insert when needed. WDYT ?

**Comment by michael-e** [comment]
Again, you lost me. It is terribly hard to follow your changes if you don't properly revert a commit before adding new stuff.

I don't understand your description either. What exactly is different from Symphony 2.6 now? I mean the behaviour, not the code — I would like to know what you want to achieve before actually looking at the code.

**Comment by nitriques** [comment]
> Again, you lost me. It is terribly hard to follow your changes if you don't properly revert a commit before adding new stuff.

Reverting would be really hard, since [I extracted part of the `add` and `edit` methods into a new method](https://github.com/symphonycms/symphony-2/commit/e99618b4a89d80dbb7b937cb46cf6b3c52dfd654#diff-57eff69d6ab1319955c5a954c701ca33R111). I've done the new commit right away because:

1. The code was *almost* the same. I went back to rev5 while git blaming the file in order to figure out why it was the way it was. I could not find an explanation.
1. I wanted to add LOCKS in the `add` method as well.

> I don't understand your description either. What exactly is different from Symphony 2.6 now?

The fact that `add` and `edit` now are sharing code. The comment should be edited to make that more clear and add a reference to the commit in question.

>  I mean the behaviour, not the code

In terms of behaviour, my goal was to have no change! As you can tell, I did not achieve the goal.

>  I would like to know what you want to achieve before actually looking at the code.

My goals with this commit are:

1. Keep the refactor, as this is a good thing.
1. Fix @pointybeard issue! By reading the code and commit comments up to rev5, it never occurred to me (or somehow I missed) the information given by @pointybeard [here](https://github.com/symphonycms/symphony-2/commit/257aab2deb688c4b6c68805b102dabd4ba880c7e#commitcomment-23450932)

> The reason this code worked (albeit a little clunky) is that it removed existing data when you want to clear out an optional field.

> By having the empty($field) check before the call to delete(), you short circuit this behaviour, meaning you can never empty out an optional field. Put another way, once you add content to an optional field, it can never be deleted, only modified to some other value.

So, my fix in commit 30d6727 restores the functionality of the optional field based on the refactored code.

In 2.6.x, both `edit` and `add` would return if field was an array. We keep this.

In 2.6.x, `add` would also return right away if the array is empty, but `edit` would not. The refactor kept the `add` behaviour (which is wrong). So the commit 30d6727 first changes that to implement what `edit` was doing (first diff block).

With this change in mind, both `edit` and `add` will try to delete existing entry data for this field. In both cases this is good. We did not issued the delete when adding stuff, but it is safe to do.

When it is time to issue the insert statement, `edit` would check that the array is empty and simply return if so. If not, it would issue the insert statement.

In 2.6.x, `add` would never have reached that point, but doing so now should not create any bugs, as long as we do not issue the insert statement. This is great, because this is exactly what `edit` needs.

**Comment by michael-e** [comment]
Thanks for this thorough explanation!

@animaux apparently already tested the fix, so if you say that you restored the "Symphony 2 behaviour" (i.e. how Symphony reacts in frontend vs. backend, which has always been different) I think that we are fine.

**Comment by nitriques** [comment]
> Thanks for this thorough explanation!

My pleasure. 

> apparently already tested the fix, so if you say that you restored the "Symphony 2 behaviour" 

For the optional fields, yes ;)

> i.e. how Symphony reacts in frontend vs. backend, which has always been different

This never intended to change, so yeah we are good. The backend will delete missing field data. The frontend won't because it populates the missing data before calling the method that would delete them.

**Comment by nilshoerrmann** [comment]
@nitriques: Is https://github.com/symphonycms/symphony-2/commit/30d67275ff20f571ab19e4089c455e6e5a415b03 the only commit I need for a quick fix?

**Comment by nitriques** [comment]
@nilshoerrmann 30d6727 is the only commit needed to fix this issue. I'll re-wrote the commit message and merge ASAP.

**Comment by nilshoerrmann** [comment]
Which branch does this commit reside in?

**Comment by nitriques** [comment]
@nilshoerrmann right now, in https://github.com/DeuxHuitHuit/symphonycms/tree/issue-2733.

**Comment by wdebusschere** [comment]
I have replaced the class.entrymanager.php with the patched one, now the optional empty field bug is gone, but when i try to empty an image field, it doesn't work 100%: in the backend I click on remove image: The file uploaded is no longer available. Please check that it exists, and is readable. On the frontend, I post an empty fields[image] nothing happens.

**Comment by nitriques** [comment]
@wdebusschere can you try the whole 2.7.x branch instead of just the entry manager ?

**Comment by nitriques** [comment]
See #2746

**Comment by brendo** [comment]
Hey @nitriques, is there an ETA on 2.7.1? It'd be great to restore this behaviour ASAP!

**Comment by nitriques** [comment]
@brendo I hope to do this week or next week.  I want this to be cleared https://github.com/symphonycms/symphony-2/issues?utf8=%E2%9C%93&q=is%3Aopen+milestone%3A2.7.1+-label%3A%22Fixed+in+2.7.x%22+
